File: 07.03.2023/py/main.py
# Write an algorithm that takes an array and moves all of the zeros to the end,
# preserving the order of the other elements.


class WrongType(Exception):
    pass

# Solution one
# time complexity of the algorithm is O(n) because it iterates over the input list once,
# and the two subsequent for-loops also iterate over the input list once. Therefore, the overall time complexity is O(n).
def move_zeros_one(lst):
    # Initialize a new list to hold the non-zero elements
    non_zero_lst = []

    # Initialize a variable to keep track of the number of zeros
    zero_count = 0

    # Iterate over the input list and append non-zero elements to the new list
    for elem in lst:
        if elem != 0:
            non_zero_lst.append(elem)
        else:
            zero_count += 1

    # Append zeros to the new list based on the zero count
    for i in range(zero_count):
        non_zero_lst.append(0)

    # Return the new list
    return non_zero_lst

# Solution two
# eliminated the second for-loop, resulting in a more efficient O(n) time complexity
def move_zeros_two(lst):
    # Initialize a variable to keep track of the last non-zero index
    last_non_zero_index = 0

    # Iterate over the input list and swap non-zero elements to the beginning of the list
    for i in range(len(lst)):
        if lst[i] != 0:
            lst[last_non_zero_index], lst[i] = lst[i], lst[last_non_zero_index]
            last_non_zero_index += 1

    # Return the modified list
    return lst


import cProfile
import time
from timeit import default_timer as timer


if __name__ == '__main__':
    cProfile.run('move_zeros_one([9, 0, 0, 9, 1, 2, 0, 1, 0, 1, 0, 3, 0, 1, 9, 0, 0, 0, 0, 9])')
    cProfile.run('move_zeros_two([9, 0, 0, 9, 1, 2, 0, 1, 0, 1, 0, 3, 0, 1, 9, 0, 0, 0, 0, 9])')

    start_time = time.time()
    move_zeros_one([9, 0, 0, 9, 1, 2, 0, 1, 0, 1, 0, 3, 0, 1, 9, 0, 0, 0, 0, 9])
    move_zeros_two([9, 0, 0, 9, 1, 2, 0, 1, 0, 1, 0, 3, 0, 1, 9, 0, 0, 0, 0, 9])
    end_time = time.time()

    print('', end_time - start_time)

File: 07.03.2023/js/main.js
// Write an algorithm that takes an array and moves all of the zeros to the end,
// preserving the order of the other elements.


/* Solution one */
// time complexity of the algorithm is O(n), where n is the length of the input array.
// This is because we need to iterate over the input array once to separate the zeros from the non-zeros,
// and then concatenate the two arrays, which takes O(n) time.
export const moveZerosOne = (arr) => {
    let zeros = [];
    let nonZeros = [];

    // Separate zeros from non-zeros
    for (let i = 0; i < arr.length; i++) {
        if (arr[i] === 0) {
            zeros.push(arr[i]);
        } else {
            nonZeros.push(arr[i]);
        }
    }

    // Concatenate non-zeros and zeros
    return nonZeros.concat(zeros);
}

/* Solution one */
// time complexity of the algorithm is O(n)
// improved the performance of this algorithm by doing the separation and concatenation in a single loop
export const moveZerosTwo = (arr) => {
    let j = 0; // index of first zero

    // Iterate through the array, swapping non-zeros with zeros
    for (let i = 0; i < arr.length; i++) {
        if (arr[i] !== 0) {
            // Swap non-zero with first zero
            let temp = arr[j];

            arr[j] = arr[i];
            arr[i] = temp;
            j++;
        }
    }

    return arr;
}
